Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -98,6 +99,92 @@ static ErrorOr<StartOfTilePart> read_start_of_tile_part(ReadonlyBytes data)
     return sot;
 }
 
+// A.5.1 Image and tile size (SIZ)
+struct ImageAndTileSize {
+    // "Denotes capabilities that a decoder needs to properly decode the codestream."
+    u16 needed_decoder_capabilities { 0 }; // "Rsiz" in spec.
+
+    // "Width of the reference grid."
+    u32 width { 0 }; // "Xsiz" in spec.
+
+    // "Height of the reference grid."
+    u32 height { 0 }; // "Ysiz" in spec.
+
+    // "Horizontal offset from the origin of the reference grid to the left side of the image area."
+    u32 x_offset { 0 }; // "XOsiz" in spec.
+
+    // "Vertical offset from the origin of the reference grid to the top side of the image area."
+    u32 y_offset { 0 }; // "YOsiz" in spec.
+
+    // "Width of one reference tile with respect to the reference grid."
+    u32 tile_width { 0 }; // "XTsiz" in spec.
+
+    // "Height of one reference tile with respect to the reference grid."
+    u32 tile_height { 0 }; // "YTsiz" in spec.
+
+    // "Horizontal offset from the origin of the reference grid to the left side of the first tile."
+    u32 tile_x_offset { 0 }; // "XTOsiz" in spec.
+
+    // "Vertical offset from the origin of the reference grid to the top side of the first tile."
+    u32 tile_y_offset { 0 }; // "YTOsiz" in spec.
+
+    // "Csiz" isn't stored in this struct. It corresponds to `components.size()`.
+
+    struct ComponentInformation {
+        // "Precision (depth) in bits and sign of the ith component samples."
+        u8 depth_and_sign { 0 }; // "Ssiz" in spec.
+
+        // Table A.11 – Component Ssiz parameter
+        u8 bit_depth() const { return (depth_and_sign & 0x7F) + 1; }
+        bool is_signed() const { return depth_and_sign & 0x80; }
+
+        // "Horizontal separation of a sample of the ith component with respect to the reference grid."
+        u8 horizontal_separation { 0 }; // "XRsiz" in spec.
+
+        // "Vertical separation of a sample of the ith component with respect to the reference grid."
+        u8 vertical_separation { 0 }; // "YRsiz" in spec.
+    };
+    Vector<ComponentInformation> components;
+};
+
+static ErrorOr<ImageAndTileSize> read_image_and_tile_size(ReadonlyBytes data)
+{
+    if (data.size() < 36)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Not enough data for SIZ marker segment");
+
+    ImageAndTileSize siz;
+    siz.needed_decoder_capabilities = *reinterpret_cast<AK::BigEndian<u16> const*>(data.data());
```

Comment:
Seems like this is how all of JPEG2000Loader.cpp is structured, so just kinda thinking out loud, but this might be somewhat easier to maintain with streams?
```c++
FixedMemoryStream stream { data };

ImageAndTileSize siz;
siz.needed_decoder_capabilities = TRY(stream.read_value<BigEndian<u16>>());
siz.width = TRY(stream.read_value<BigEndian<u32>>());
```

Seems like it'd be less manual size / offset computation, but I don't feel strongly about it.